### ISSUE TITLE
Allow global shell commands/custom commands to be run from everywhere (not just inside a project)

### ISSUE THREAD

**Issue by OlivierSC** [ddev/ddev#4416]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

As it's possible to start a specific project from everywhere on host with :
From my home folder ~/
`ddev start projectname`
-> projectname starts

Currently, custom commands seems to need to already be in a project folder to run.
If run from outside existing ddev project folder, global commands (from ~/.ddev/commands/host ) are not found : 
> unknown command "phpstorm" for "ddev"

it could be very useful to run command (ex with "cd" on linux/mac to go to folder) on the specified host project folder from everywhere.
example : 
I'm in my home folder ~/
`ddev cd projectname`
-> open the projectname folder path.

It could be also usefull for existing command  :
I'm in my home folder ~/
`ddev phpstorme projectname`
-> open the projectname project in phpstorm.

This could be very useful when working with lots of projects rather than manually change folder / open project.

### Describe your solution

The first step would be to allow commands to be run from everywhere, even outside from project folders (ex my "home" folder).

The second step would be to find the best way to provide the "project name" parameter to the command, even for existing command like phpstorm.

A bonus step would be that ddev could try to find local commands ([project folder/.ddev/commands) for the specified "project name", that way not all commands need to be "globals".

**Comment by chrisfromredfin** [comment]
There's also the idea that some host commands might not even need a project, for example the new self-upgrade. It would be nice if that (or ones like it) could run without a project context, and without needing to start one project first.

**Comment by rfay** [comment]
I've been thinking about this a bit and it has potential.

1. Instead of copying global commands into project .global_commands in-container commands should be copied into /mnt/ddev-global-config
2. host global commands can be used in place, and should be available from anywhere.
3. Commands themselves will have to determine whether they require project context or not, and maybe not show. I think perhaps an annotation for that would be an elegant way to do it.

**Comment by GuySartorelli** [comment]
I'm keen for this to be implemented. @rfay is https://github.com/ddev/ddev/issues/4416#issuecomment-1838665202 a full set of acceptance criteria for this feature?
If not, what other criteria would you like for this?

I'd be keen to help get this implemented.

**Comment by rfay** [comment]
It's good enough for a first draft, we may encounter complexities.

**Comment by GuySartorelli** [comment]
I'm going to tackle this in a series (probably just two) of PRs, since there are a nice set of distinct pieces of work this can be split into.

I've raised https://github.com/ddev/ddev/pull/5702 so far, which covers 2 and 3 from https://github.com/ddev/ddev/issues/4416#issuecomment-1838665202

**Comment by GuySartorelli** [comment]
I'd also like to look into adding a "project name" argument that, if used, passes the relevant details for that project to the command. This was mentioned in the issue description and I think it'd be really valuable to be able to, for example, run `ddev launch --project-name my-project` and have it execute in the context of that project.

My current thinking to do that in a way that's backwards compatible is one of:
1. Any flag in the "Flags" annotation called `--project-name` is assumed to take the name of a DDEV project, and activates this new logic, OR
1. Add a special key for flags in the "Flags" annotation that is used to identify which flag (if any) should activate this logic.

1 above more consistent, but could cause problems for existing projects with a `--project-name` flag that _should not_ activate this new logic.
2 is fully backwards compatible since it's 100% opt-in, but it allows developers to have a different name for the flag they want the project name to be passed to, which gives a less consistent user experience.

@rfay do you have any thoughts on this?

**Comment by rfay** [comment]
Launch is a strange fellow, and there have been a number of recent conversations about it. I don't see why `--project-name` shouldn't work, although I think `--project` might be simpler and easier.

I don't think that `--project-name` should cause an error if used in a different project.

It's possible that `launch` just does too much and should be re-thought.

**Comment by GuySartorelli** [comment]
`launch` was just an example - there are plenty of 9ther commands that could also very easily be slightly reworked to use a flag. The idea is that it would be implemented in a generic way that any custom global host command can use it.

**Comment by GuySartorelli** [comment]
So I guess my questions for that are:
1. Is that something you'd be keen to support if I implemented it?
2. Would you like it based on the name of the flag, or on some new piece of data (i.e. 1 or 2 above)?
3. If it's based on the name, what should the name be? Would you want to change any existing flags that serve this purpose to match this new name?
4. What would your expected behaviour be when this flag is used and the user's cwd is inside a project?

I'm keen to implement this once #4416 is merged if it's something you'd support.

**Comment by GuySartorelli** [comment]
@rfay For this:
> Instead of copying global commands into project .global_commands in-container commands should be copied into /mnt/ddev-global-config

The directory path `/mnt/ddev-global-config` looks like what I'd expect for a volume - but I'm not sure what you'd expect the volume to map to on the host?

But then the word "copied" there suggests it's probably not a volume, because if it was a volume you wouldn't need to copy them in?

I'm a bit confused about the intention here. Seems to me we should probably do one of:

1. Have a `/mnt/commands` that mounts from the specific commands directory for that service. For example the web service would mount `~/.ddev/commands/web/` onto `/mnt/commands`. Nothing needs to be copied - it's just there.
    The downsides of this approach are deleting/editing files in the service will affect the files on the host, and each service needs to be declared slightly differently from one another. This would be a breaking change, because no custom services will have this volume right now.
1. Don't use a volume at all. Just have a directory on the container like `/etc/ddev/commands/` (or wherever else you want it) which the commands for that service get _copied into_ (like how they're currently copied into the project's `.ddev/.global_commands/` directory).
    This doesn't have any of the downsides of option 1. Services don't have to define anything special in their docker compose yaml (which is actually an improvement on the current status quo).

For now I'll start working on option 2 above, since that seems to be the way to go - but if I've misunderstood something or you want something different done just let me know.

**Comment by GuySartorelli** [comment]
I've started work on that in https://github.com/ddev/ddev/pull/5753

**Comment by rfay** [comment]
ddev-global-cache is indeed a volume currently used for these things (had a typo above, working on phone in a campground :) ) I would have been expecting /mnt/ddev-global-cache/

It's mounted into web and db containers and usually everywhere else. So the idea works for custom commands that are *not* host commands. 

It does not work for host commands, but we should be able to find a way to run project host commands in place and also global host commands in place.

**Comment by GuySartorelli** [comment]
Ahh okay that makes more sense. I did see that mount and was going to use it as a template for how to set up the new one - that solves that problem for me haha.

> It does not work for host commands, but we should be able to find a way to run project host commands in place and also global host commands in place.

Host commands don't need to be in containers anyway, so that's fine. Global host commands will stay where they already are (usually `~/.ddev/commands/host`) and be run from there. That's already a solved problem in my POC PR.

**Comment by GuySartorelli** [comment]
@rfay I was about to start working on changing my POC PR to use the `/mnt/ddev-global-cache/` mount, but I realised very quickly that (as its name suggests) it is a _global_ cache.
That means if I wanted to use it, the commands for _every project_ would end up in there - including any which are specific to that project, which could very easily cause collisions.

We'd need to have nested directories for each project to avoid collisions, something like either `commands/<service>/<projectname>/` or `commands/<projectname>/<service>/` - either way, it seems like this probably isn't the right place for these. Global cache seems like where you'd want shared stuff but project commands are explicitly not intended to be shared.

What's more, since the mount isn't explicitly tied to either the containers directly or the project directory, if someone incorrectly removes their project (e.g. `rm -r <project-dir>` and then `docker rm ...` to remove the container), that project's commands will still be sitting there (which is a problem the current setup avoids).

**Comment by rfay** [comment]
Take a look at what's in there, there are lots of usages that are keyed by container name. `ddev ssh` and visit /mnt/ddev-global-cache, look at `bashhistory` for example, there's a directory for each container that has been run. `nvm_dir` is the same. `composer` is not done that way because it sorts its own stuff.

**Comment by GuySartorelli** [comment]
Ahhh okay. Well if that's the precedent already I don't have any qualms about it then.
I'll add something to `ddev clean` to also remove commands from that directory - should `ddev delete` _also_ delete them? I tend to use `ddev delete` whenever I am ready to get rid of a project, but I almost never use `ddev clean`.

**Comment by rfay** [comment]
Yes, `ddev clean` and `ddev delete` should clean up things like this, as they do for the other things. Hopefully.

**Comment by GuySartorelli** [comment]
This was resolved in https://github.com/ddev/ddev/pull/5753
The only outstanding piece is having a flag that lets you run project commands from arbitrary paths, which is essentially just a duplicate of https://github.com/ddev/ddev/issues/2128

**Comment by rfay** [comment]
Thanks for the great work on this!
